### PR TITLE
[Feature] add exchange_bytes() and exchange_speed() to profile exchange operator

### PR DESCRIPTION
**Pull request by fzhedu**
Signed-off-by: fzhedu <fzhedu@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #
```
mysql> explain select host_name(), exchange_speed(lo_orderpriority,lo_shipmode) from lineorder;
ERROR 1064 (HY000): exchange_speed should run in new_planner_agg_stage = 1.
mysql> set new_planner_agg_stage = 1;
Query OK, 0 rows affected (0.02 sec)

mysql>  select host_name(), exchange_speed(l_comment) from lineitem group by l_partkey;
+-------------+-----------------------------------------------------------------+
| host_name() | exchange_speed(l_comment)                                       |
+-------------+-----------------------------------------------------------------+
| pipeline04  | exchange 6100246805 bytes in 21.655532 s, speed = 268.6449 MB/s |
| pipeline06  | exchange 6101295089 bytes in 21.655697 s, speed = 268.6890 MB/s |
| pipeline05  | exchange 6100219633 bytes in 21.654098 s, speed = 268.6615 MB/s |
+-------------+-----------------------------------------------------------------+

mysql> select exchange_speed(l_orderkey),exchange_bytes(l_orderkey) from lineitem;
+--------------------------------------------------------------+----------------------------+
| exchange_speed(l_orderkey)                                   | exchange_bytes(l_orderkey) |
+--------------------------------------------------------------+----------------------------+
| exchange 2400151608 bytes in 1.863807 s, speed = 1.1993 GB/s |                 2400151608 |
+--------------------------------------------------------------+----------------------------+
1 row in set (1.89 sec)

```

`set enable_exchange_perf` helps to ignore local data in order to test the real network speed during exchange.

### Idea:

like one-phase aggregation, repartitioning or collecting data among nodes via exchange, then using exchange exchange_bytes() and exchange_speed() to profile exchange.

### NOTE:
should set the following two parameters first
```
set new_planner_agg_stage = 1;
set enable_exchange_perf = true;
```

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] I have added user [document](https://starrocks.feishu.cn/docx/doxcnofGoH3XB9YV74TSxCOAg0c) for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 44 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[662, 664]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[597, 598, 600]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|8|75.00%|[66, 131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|4|75.00%|[364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|6|83.33%|[166]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 44 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[662, 664]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[597, 598, 600]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|8|75.00%|[66, 131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|4|75.00%|[364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|6|83.33%|[166]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 44 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[662, 664]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[597, 598, 600]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|8|75.00%|[66, 131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|4|75.00%|[364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|6|83.33%|[166]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 44 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[662, 664]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[597, 598, 600]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|8|75.00%|[66, 131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|4|75.00%|[364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|6|83.33%|[166]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 45 / 56 (80.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[662, 664]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[597, 598, 600]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|8|75.00%|[66, 131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|4|75.00%|[364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|6|83.33%|[166]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|15|15|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/EnforceAndCostTask.java [hunk 1]
```diff
@@ -356,6 +357,12 @@ private boolean canGenerateOneStageAgg(GroupExpression childBestExpr) {
         }
 
         PhysicalHashAggregateOperator aggregate = (PhysicalHashAggregateOperator) groupExpression.getOp();
+        // Must do one stage aggregate If there are exchange_bytes/_speed()
+        if (aggregate.getAggregations().values().stream().anyMatch(aggFunc ->
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/EnforceAndCostTask.java:361]: why must do one stage aggregate?

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java [hunk 2]
```diff
@@ -589,6 +592,15 @@ public Function getFunction(Function desc, Function.CompareMode mode) {
             return null;
         }
 
+        // specially matching exactly one function by name
+        if (mode == Function.CompareMode.MATCH_NAME) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:596]: why need this logic?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java [hunk 3]
```diff
@@ -53,7 +53,10 @@ public ScalarOperator visitCall(CallOperator call, ScalarOperatorRewriteContext
                 }
             }
         } else {
-            if (fn.functionName().equals(FunctionSet.ARRAY_MAP)) { // array_map does not need to implicit cast.
+            // functions with various data types do not need to implicit cast, such as following functions.
+            if (fn.functionName().equals(FunctionSet.ARRAY_MAP) ||
+                    fn.functionName().equals(FunctionSet.EXCHANGE_BYTES) ||
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java:58]: why disable implicit cast for these new exchange functions?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/EnforceAndCostTask.java (hunk 1 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/EnforceAndCostTask.java:361]: if using two stage, the exchange data becomes smaller, which is not helpful to perf exchange.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java (hunk 3 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java:58]: these functions accept any data type parameter, it is no need to cast to a destnation type.

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java (hunk 2 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:596]: previous logic does not consider functions with various parameters and any data types, these functions just need to match the function name.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 44 / 55 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[650, 652]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[598, 599, 601]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|6|8|75.00%|[66, 131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|3|4|75.00%|[364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|6|83.33%|[166]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 30 / 37 (81.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[650, 652]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[598, 599, 601]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 28 / 39 (71.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|2|6|33.33%|[324, 325, 327, 328]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|3|33.33%|[650, 652]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|2|4|50.00%|[224, 225]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|7|10|70.00%|[598, 599, 601]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 33 (78.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|4|25.00%|[650, 652, 653]|
|:large_blue_circle:|com/starrocks/catalog/Function.java|1|3|33.33%|[221, 222]|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|2|4|50.00%|[324, 325]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 25 / 32 (78.12%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Function.java|0|2|00.00%|[221, 222]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|4|25.00%|[650, 652, 653]|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|2|4|50.00%|[324, 325]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 4]
```diff
@@ -1205,18 +1205,25 @@ public PlanFragment visitPhysicalHashAggregate(OptExpression optExpr, ExecPlan c
             TupleDescriptor outputTupleDesc = context.getDescTbl().createTupleDescriptor();
 
             ArrayList<Expr> groupingExpressions = Lists.newArrayList();
+            // EXCHANGE_BYTES/_SPEED aggregate the total bytes/ratio on a node, without grouping, remove group-by here.
+            // the group-by expressions just denote the hash distribution of an exchange operator.
+            boolean forExchange = node.getAggregations().values().stream().anyMatch(aggFunc ->
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1210]: would better give a better name to `forExchange`

Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java]: changed to forExchangePerf

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 30 / 37 (81.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Function.java|0|2|00.00%|[221, 222]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|4|25.00%|[650, 652, 653]|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|2|4|50.00%|[324, 325]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|15|15|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**